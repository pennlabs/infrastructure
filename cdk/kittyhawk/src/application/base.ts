import { Construct } from "constructs";
import { removeSubdomain } from "..";
import { Certificate } from "../certificate";
import { Deployment, DeploymentProps } from "../deployment";
import { Ingress, IngressProps } from "../ingress";
import { Service } from "../service";
import { ServiceAccount } from "../serviceaccount";

export interface ApplicationProps {
  /**
   * Ingress configuration
   */
  readonly ingress?: IngressProps;

  /**
   * Deployment configuration
   */
  readonly deployment: DeploymentProps;

  /**
   * Port to expose the application on.
   */
  readonly port?: number;

  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}

export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || "undefined_release";
    const fullname = `${release_name}-${appname}`;

    new Service(this, fullname, props.port);

    if (props.createServiceAccount) {
      new ServiceAccount(this, `${appname}-${release_name}`, {
        serviceAccountName: release_name,
      });
    }

    new Deployment(this, fullname, {
      ...props.deployment,
      port: props.port,
      ...(props.createServiceAccount
        ? { serviceAccountName: release_name }
        : {}),
    });

    if (props.ingress) {
      new Ingress(this, fullname, props.ingress);

      const alreadyCreatedCertificates = new Set<string>();

      props.ingress.rules.forEach((rules) => {
        const finalDomain: string = removeSubdomain(
          rules.host,
          rules.isSubdomain ?? false
        );
        if (!alreadyCreatedCertificates.has(finalDomain)) {
          new Certificate(this, fullname, rules);
          alreadyCreatedCertificates.add(finalDomain);
        }
      });
    }
  }
}
