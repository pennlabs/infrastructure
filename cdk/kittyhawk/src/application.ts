import { Construct } from 'constructs';
import { Deployment, DeploymentProps } from './deployment';
import { Ingress, IngressProps } from './ingress';
import { Service, ServiceProps } from './service'
import { Certificate } from './certificate';

/**
 * Warning: Before editing any interfaces, make sure that none of the interfaces will have 
 * property names that conflict with each other. Typescript may not throw an error and it 
 * could cause problems.
 */
export interface ApplicationProps extends IngressProps, DeploymentProps,
  ServiceProps { }


export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || 'undefined_release';
    const fullname = `${release_name}-${appname}`

    new Service(this, fullname, props);

    new Deployment(this, fullname, props);

    if (props.ingress) {
      new Ingress(this, fullname, props)

      new Certificate(this, fullname, props)
    }
  }
}

/**
 * Helper function that first checks to make sure that the environment variable array 
 * doesn't already contain the env var, then inserts it into the array.
 * @param envArray array of environment variables to insert into 
 * @param envKey name of the environment variable to insert
 * @param envValue value of the environment variable
 */
function insertIfNotPresent(envArray: { name: string, value: string }[], envKey: string, envValue: any) {
  const envSettingsModule = envArray?.filter(env => (env.name === envKey));
  if (envSettingsModule?.length > 0) {
    throw new Error(`${envKey} should not be redefined as an enviroment variable.`)
  }
  envArray.push({ name: envKey, value: envValue });

}
export interface DjangoApplicationProps extends ApplicationProps {
  /**
   * Domain of the application.
   */
  readonly domain: string;

  /**
   * Just the list of paths passed to the ingress since we already know the host.
   */
  readonly ingressPaths: string[];

  /**
   * DJANGO_SETTINGS_MODULE environment variable.
   */
  readonly djangoSettingsModule: string;
}

export class DjangoApplication extends Application {
  constructor(scope: Construct, appname: string, props: DjangoApplicationProps) {

    // Have to be careful here with references when mutating things
    let djangoExtraEnv = Array.from(props.extraEnv || []);

    // Insert DJANGO_SETTINGS_MODULE and DOMAIN
    insertIfNotPresent(djangoExtraEnv, 'DJANGO_SETTINGS_MODULE', props.djangoSettingsModule)
    insertIfNotPresent(djangoExtraEnv, 'DOMAIN', props.domain)

    // Configure the ingress using ingressPaths.
    const djangoIngress = [{ host: props.domain, paths: props.ingressPaths }]

    // If everything passes, construct the Application.
    super(scope, appname, {
      ...props,
      extraEnv: djangoExtraEnv,
      ingress: djangoIngress,
    });
  }
}

export interface ReactApplicationProps extends ApplicationProps {
  /**
   * Domain of the application.
   */
  readonly domain: string;

  /**
   * Just the list of paths passed to the ingress since we already know the host.
   */
  readonly ingressPaths: string[];

  /**
   * PORT environment variable for react. Default '80'.
   */
  readonly portEnv?: string;
}

export class ReactApplication extends Application {
  constructor(scope: Construct, appname: string, props: ReactApplicationProps) {

    // Have to be careful here with references when mutating things
    let reactExtraEnv = Array.from(props.extraEnv || []);

    // Insert DOMAIN and PORT as env vars.
    insertIfNotPresent(reactExtraEnv, 'DOMAIN', props.domain)
    insertIfNotPresent(reactExtraEnv, 'PORT', props.portEnv || '80')

    // Configure the ingress using ingressPaths.
    const reactIngress = [{ host: props.domain, paths: props.ingressPaths }]

    // If everything passes, construct the Application.
    super(scope, appname, {
      ...props,
      extraEnv: reactExtraEnv,
      ingress: reactIngress,
    });
  }
}

export class RedisApplication extends Application {
  constructor(scope: Construct, appname: string, redisProps: Partial<ApplicationProps>) {
    super(scope, appname, {
      ...redisProps,
      image: 'redis',
      tag: redisProps.tag || '6.0',
      port: redisProps.port || 6379,
    });
  }
}
