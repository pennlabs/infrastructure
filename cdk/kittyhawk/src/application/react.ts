import { Construct } from "constructs";
import { DeploymentProps } from "../deployment";
import { HostRules, IngressProps } from "../ingress";
import { Application } from "./base";

export interface ReactApplicationProps {
  /**
   * Deployment configuration
   */
  readonly deployment: DeploymentProps;

  /**
   * Port to expose the application on.
   */
  readonly port?: number;

  /**
   * Host is the domain the application runs on,
   * isSubdomain is true if the domain should be treated as a subdomain for certificate purposes.
   * paths is the list of paths to expose the application on.
   * See the certificate documentation for more details.
   *
   */
  readonly domain: HostRules;

  /**
   * Optional ingressProps to override the default ingress props.
   */
  readonly ingressProps?: IngressProps;

  /**
   * PORT environment variable for react. Default '80'.
   */
  readonly portEnv?: string;

  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}

export class ReactApplication extends Application {
  constructor(scope: Construct, appname: string, props: ReactApplicationProps) {
    // Now, we ensure there are no duplicate env variables, even if they redefine it
    const reactExtraEnv = [
      ...new Set([
        ...(props.deployment?.env || []),
        { name: "DOMAIN", value: props.domain.host },
        { name: "PORT", value: props.portEnv || "80" },
      ]),
    ];

    // If everything passes, construct the Application.
    super(scope, appname, {
      port: props.port ?? 80,
      deployment: {
        ...props.deployment,
        env: reactExtraEnv,
      },
      ingress: {
        rules: [props.domain],
        ...props.ingressProps,
      },
      createServiceAccount: props.createServiceAccount,
    });
  }
}
