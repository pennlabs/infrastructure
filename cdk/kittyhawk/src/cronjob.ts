import { Construct } from 'constructs';
import { KubeCronJobV1Beta1 as CronJobApiObject } from '../imports/k8s';
import { Container, ContainerProps, Volume } from './container';

export interface CronJobProps extends ContainerProps {

  /**
     * The schedule in Cron format.
     */
  readonly schedule: string;

  /**
    * Restart policy for all containers.
    * 
    * @default "Never"
    * 
    */

  readonly restartPolicy?: 'Always' | 'OnFailure' | 'Never';

  /**
    * The number of successful finished jobs to retain. 
    * 
    * @default 1
    * 
    */
  readonly successLimit?: number;

  /**
    * The number of failed jobs to retain.
    * 
    * @default 1
    * 
    */
  readonly failureLimit?: number;

  /**
    * Secret volume mounts for cronjob container.
    *
    * @default undefined
    */
  readonly secretMounts?: { name: string, mountPath: string, subPath: string }[]

}


export class CronJob extends Construct {
  constructor(scope: Construct, jobname: string, props: CronJobProps) {
    super(scope, jobname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || 'undefined_release';
    const fullname = `${release_name}-${jobname}`
    const containers: Container[] = [new Container(props)];
    const volumes: Volume[] | undefined = props.secretMounts?.map(m => new Volume(m))

    new CronJobApiObject(this, `cronjob-${fullname}`, {
      metadata: {
        name: fullname,
        namespace: 'default',
        labels: { name: fullname },
      },
      spec: {
        schedule: props.schedule,
        jobTemplate: {
          spec: {
            template: {
              spec: {
                restartPolicy: props.restartPolicy || 'Never',
                containers: containers,
                volumes: volumes,
              },
            },

          },
        },
        failedJobsHistoryLimit: props.failureLimit ?? 1,
        successfulJobsHistoryLimit: props.successLimit ?? 1,
      },
    });
  }
}