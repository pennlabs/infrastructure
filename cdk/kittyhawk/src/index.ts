
export * from './application';
export * from './autoscaler';
export * from './certificate';
export * from './chart';
export * from './container';
export * from './cronjob';
export * from './deployment';
export * from './ingress';
export * from './service';
