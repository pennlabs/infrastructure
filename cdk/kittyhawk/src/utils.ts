export type NonEmptyArray<T> = [T, ...T[]];

export function nonEmptyMap<T, U>(
  arr: NonEmptyArray<T>,
  callbackfn: (value: T) => U
): NonEmptyArray<U> {
  return arr.map(callbackfn) as any as NonEmptyArray<U>;
}
