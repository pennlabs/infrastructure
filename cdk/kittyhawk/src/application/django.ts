import { Construct } from 'constructs';
import { Application } from '.';
import { DeploymentProps } from '../deployment';
import { HostRules, IngressProps } from '../ingress';
import { NonEmptyArray, nonEmptyMap } from '../utils';

export interface DjangoApplicationProps {
  /**
   * Deployment configuration
   */
  readonly deployment: DeploymentProps;

  /**
   * Port to expose the application on.
   */
  readonly port?: number;

  /**
   * Array of domain(s). Host is the domain the application runs on,
   * and isSubdomain is true if the domain should be treated as a subdomain for certificate purposes.
   * See the certificate documentation for more details.
   */
  readonly domains: NonEmptyArray<{ host: string; isSubdomain?: boolean }>;

  /**
   * Just the list of paths passed to the ingress since we already know the host. Optional.
   *
   * @default undefined
   */
  readonly ingressPaths?: string[];

  /**
   * Optional ingressProps to override the default ingress props.
   */
  readonly ingressProps?: IngressProps;

  /**
   * DJANGO_SETTINGS_MODULE environment variable.
   */
  readonly djangoSettingsModule: string;

  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}


export default class DjangoApplication extends Application {
  constructor(scope: Construct, appname: string, props: DjangoApplicationProps) {

    // Now, we ensure there are no duplicate env variables, even if they redefine it
    const djangoExtraEnv = [...new Set([
      ...props.deployment?.env || [],
      { name: 'DJANGO_SETTINGS_MODULE', value: props.djangoSettingsModule },
      { name: 'DOMAIN', value: nonEmptyMap(props.domains, (h => h.host)).join() },
    ])];

    // Configure the ingress using ingressPaths if ingressPaths is defined.
    const djangoIngress: HostRules[] = props.domains?.map(h => ({
      host: h.host,
      paths: props.ingressPaths || [],
      isSubdomain: h.isSubdomain ?? false,
    }));

    // If everything passes, construct the Application.
    super(scope, appname, {
      port: props.port ?? 80,
      deployment: {
        ...props.deployment,
        env: djangoExtraEnv,
      },
      ingress: { 
        rules: djangoIngress as NonEmptyArray<HostRules>,
        ...props.ingressProps,
       },
      createServiceAccount: props.createServiceAccount,
    });
  }
}