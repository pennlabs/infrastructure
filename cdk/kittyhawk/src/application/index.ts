import { Construct } from 'constructs';
import { Certificate } from '../certificate';
import { Deployment, DeploymentProps } from '../deployment';
import { Ingress, IngressProps } from '../ingress';
import { Service } from '../service';
import { ServiceAccount } from '../serviceaccount';
import { nonEmptyMap } from '../utils';

import DjangoApplication from './django';
import ReactApplication from './react';
import RedisApplication from './redis';

export { DjangoApplication };
export { ReactApplication };
export { RedisApplication };

/**
 * Warning: Before editing any interfaces, make sure that none of the interfaces will have
 * property names that conflict with each other. Typescript may not throw an error and it
 * could cause problems.
 */
export interface ApplicationProps {
  readonly ingress?: IngressProps;
  readonly deployment: DeploymentProps;
  readonly port?: number;
  /**
   * Creates a service account and attach it to any deployment pods.
   * serviceAccountName: release name
   */
  readonly createServiceAccount?: boolean;
}

export class Application extends Construct {
  constructor(scope: Construct, appname: string, props: ApplicationProps) {
    super(scope, appname);

    // We want to prepend the project name to the name of each component
    const release_name = process.env.RELEASE_NAME || 'undefined_release';
    const fullname = `${release_name}-${appname}`;

    new Service(this, fullname, props.port);

    if (props.createServiceAccount) {
      new ServiceAccount(this, `${appname}-${release_name}`, {
        serviceAccountName: release_name
      })
    }

    new Deployment(this, fullname, {
      ...props.deployment,
      ...(props.createServiceAccount ? { serviceAccountName: release_name } : {}),
    });

    if (props.ingress) {
      new Ingress(this, fullname, props.ingress);

      nonEmptyMap(props.ingress.rules, rule => {
        return new Certificate(this, fullname, rule);
      });
    }
  }
}