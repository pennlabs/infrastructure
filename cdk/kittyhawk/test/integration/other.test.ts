import { Construct } from 'constructs';
import { CronJob } from '../../src/cronjob';
import { Application, DjangoApplication, ReactApplication, RedisApplication } from '../../src/application'
import { chartTest } from '../utils'
import cronTime from 'cron-time-generator';

export function buildWebsiteChart(scope: Construct) {

  /** Penn Labs Website
   * https://github.com/pennlabs/website/blob/master/k8s/values.yaml
   */
  new ReactApplication(scope, 'serve', {
    image: 'pennlabs/website',
    domain: 'pennlabs.org',
    ingressPaths: ['/'],
  })
}

export function buildBasicsChart(scope: Construct) {

  /** Penn Basics 
   * https://github.com/pennlabs/penn-basics/blob/master/k8s/values.yaml
  */
  new ReactApplication(scope, 'react', {
    image: 'pennlabs/penn-basics',
    secret: 'penn-basics',
    portEnv: '80',
    domain: 'pennbasics.com',
    ingressPaths: ['/'],
  })
}


export function buildPlatformChart(scope: Construct) {

  /** Platform
   * https://github.com/pennlabs/platform/blob/master/k8s/values.yaml
   */

  const common = {
    image: 'pennlabs/platform',
    secret: 'platform',
  }

  new DjangoApplication(scope, 'platform', {
    ...common,
    port: 443,
    domain: 'platform.pennlabs.org',
    djangoSettingsModule: 'Platform.settings.production',
    ingressPaths: ['/'],
    secretMounts: [{ name: 'platform', subPath: 'SHIBBOLETH_CERT', mountPath: '/etc/shibboleth/sp-cert.pem' },
      { name: 'platform', subPath: 'SHIBBOLETH_KEY', mountPath: '/etc/shibboleth/sp-key.pem' }],
  })

  new CronJob(scope, 'clear-expired-tokens', {
    ...common,
    schedule: cronTime.everySundayAt(5),
    cmd: ['python3', 'manage.py', 'cleartokens'],
  })
}


export function buildCFAChart(scope: Construct) {

  /** Common Funding Application
   * https://github.com/pennlabs/common-funding-application/blob/master/k8s/values.yaml
   */
  new DjangoApplication(scope, 'django', {
    image: 'pennlabs/common-funding-application',
    secret: 'common-funding-application',
    domain: 'penncfa.com',
    ingressPaths: ['/'],
    djangoSettingsModule: 'penncfa.settings.production',
  })
}


export function buildLabsAPIServerChart(scope: Construct) {
  /**
   * Labs API Server
   * https://github.com/pennlabs/labs-api-server/blob/master/k8s/values.yaml
   */
  const common = {
    image: 'pennlabs/labs-api-server',
    secret: 'labs-api-server',
  }
  new Application(scope, 'flask', {
    ...common,
    ingress: [{ host: 'api.pennlabs.org', paths: ['/'] }],
    secretMounts: [{ name: 'labs-api-server', subPath: 'ios-key', mountPath: '/app/ios_key.p8' }],
  })

  new RedisApplication(scope, 'redis', { tag: '5' })

  new CronJob(scope, 'laundry', {
    ...common,
    schedule: cronTime.every(15).minutes(),
    cmd: ['python3', 'cron/save_laundry_data.py'],
  })

  new CronJob(scope, 'gsr-notifications', {
    ...common,
    schedule: '20,50 * * * *', // Not sure how to write this with cronTime
    cmd: ['python3', 'cron/send_gsr_push_notifications.py'],
    secretMounts: [{ name: 'labs-api-server', subPath: 'ios-key', mountPath: '/app/ios_key.p8' }],
  })
}

test('Penn Labs Website', () => chartTest(buildWebsiteChart));
test('Penn Basics', () => chartTest(buildBasicsChart));
test('Platform API', () => chartTest(buildPlatformChart));
test('CFA', () => chartTest(buildCFAChart))

