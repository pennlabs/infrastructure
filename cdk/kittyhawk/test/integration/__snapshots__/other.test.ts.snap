// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CFA 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-django",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-django",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-django",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-django",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-django",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-django",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "DJANGO_SETTINGS_MODULE",
                  "value": "penncfa.settings.production",
                },
                Object {
                  "name": "DOMAIN",
                  "value": "penncfa.com",
                },
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "envFrom": Array [
                Object {
                  "secretRef": Object {
                    "name": "common-funding-application",
                  },
                },
              ],
              "image": "pennlabs/common-funding-application:TAG_FROM_CI",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-django",
      "namespace": "default",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "penncfa.com",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-django",
                  "servicePort": 80,
                },
                "path": "/",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "penncfa.com",
          ],
          "secretName": "penncfa-com-tls",
        },
      ],
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "penncfa-com",
    },
    "spec": Object {
      "dnsNames": Array [
        "penncfa.com",
        "*.penncfa.com",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "penncfa-com-tls",
    },
  },
]
`;

exports[`Penn Basics 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-react",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-react",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-react",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-react",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-react",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-react",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "DOMAIN",
                  "value": "pennbasics.com",
                },
                Object {
                  "name": "PORT",
                  "value": "80",
                },
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "envFrom": Array [
                Object {
                  "secretRef": Object {
                    "name": "penn-basics",
                  },
                },
              ],
              "image": "pennlabs/penn-basics:TAG_FROM_CI",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-react",
      "namespace": "default",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "pennbasics.com",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-react",
                  "servicePort": 80,
                },
                "path": "/",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "pennbasics.com",
          ],
          "secretName": "pennbasics-com-tls",
        },
      ],
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "pennbasics-com",
    },
    "spec": Object {
      "dnsNames": Array [
        "pennbasics.com",
        "*.pennbasics.com",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "pennbasics-com-tls",
    },
  },
]
`;

exports[`Penn Labs Website 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-serve",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 80,
          "targetPort": 80,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-serve",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-serve",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-serve",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-serve",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-serve",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "DOMAIN",
                  "value": "pennlabs.org",
                },
                Object {
                  "name": "PORT",
                  "value": "80",
                },
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "image": "pennlabs/website:TAG_FROM_CI",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 80,
                },
              ],
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-serve",
      "namespace": "default",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "pennlabs.org",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-serve",
                  "servicePort": 80,
                },
                "path": "/",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "pennlabs.org",
          ],
          "secretName": "pennlabs-org-tls",
        },
      ],
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "pennlabs-org",
    },
    "spec": Object {
      "dnsNames": Array [
        "pennlabs.org",
        "*.pennlabs.org",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "pennlabs-org-tls",
    },
  },
]
`;

exports[`Platform API 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-platform",
    },
    "spec": Object {
      "ports": Array [
        Object {
          "port": 443,
          "targetPort": 443,
        },
      ],
      "selector": Object {
        "name": "RELEASE_NAME-platform",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-platform",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-platform",
      "namespace": "default",
    },
    "spec": Object {
      "replicas": 1,
      "selector": Object {
        "matchLabels": Object {
          "name": "RELEASE_NAME-platform",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": 3,
          "maxUnavailable": 0,
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "name": "RELEASE_NAME-platform",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "env": Array [
                Object {
                  "name": "DJANGO_SETTINGS_MODULE",
                  "value": "Platform.settings.production",
                },
                Object {
                  "name": "DOMAIN",
                  "value": "platform.pennlabs.org",
                },
                Object {
                  "name": "GIT_SHA",
                  "value": "TAG_FROM_CI",
                },
              ],
              "envFrom": Array [
                Object {
                  "secretRef": Object {
                    "name": "platform",
                  },
                },
              ],
              "image": "pennlabs/platform:TAG_FROM_CI",
              "imagePullPolicy": "IfNotPresent",
              "name": "worker",
              "ports": Array [
                Object {
                  "containerPort": 443,
                },
              ],
              "volumeMounts": Array [
                Object {
                  "mountPath": "/etc/shibboleth/sp-cert.pem",
                  "name": "platform",
                  "subPath": "SHIBBOLETH_CERT",
                },
                Object {
                  "mountPath": "/etc/shibboleth/sp-key.pem",
                  "name": "platform",
                  "subPath": "SHIBBOLETH_KEY",
                },
              ],
            },
          ],
          "volumes": Array [
            Object {
              "name": "platform-shibboleth-cert",
              "secret": Object {
                "items": Array [
                  Object {
                    "key": "SHIBBOLETH_CERT",
                    "path": "SHIBBOLETH_CERT",
                  },
                ],
                "secretName": "platform",
              },
            },
            Object {
              "name": "platform-shibboleth-key",
              "secret": Object {
                "items": Array [
                  Object {
                    "key": "SHIBBOLETH_KEY",
                    "path": "SHIBBOLETH_KEY",
                  },
                ],
                "secretName": "platform",
              },
            },
          ],
        },
      },
    },
  },
  Object {
    "apiVersion": "networking.k8s.io/v1beta1",
    "kind": "Ingress",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-platform",
      "namespace": "default",
    },
    "spec": Object {
      "rules": Array [
        Object {
          "host": "platform.pennlabs.org",
          "http": Object {
            "paths": Array [
              Object {
                "backend": Object {
                  "serviceName": "RELEASE_NAME-platform",
                  "servicePort": 443,
                },
                "path": "/",
              },
            ],
          },
        },
      ],
      "tls": Array [
        Object {
          "hosts": Array [
            "platform.pennlabs.org",
          ],
          "secretName": "pennlabs-org-tls",
        },
      ],
    },
  },
  Object {
    "apiVersion": "cert-manager.io/v1alpha2",
    "kind": "Certificate",
    "metadata": Object {
      "labels": Object {
        "release": "RELEASE_NAME",
      },
      "name": "pennlabs-org",
    },
    "spec": Object {
      "dnsNames": Array [
        "pennlabs.org",
        "*.pennlabs.org",
      ],
      "issuerRef": Object {
        "group": "cert-manager.io",
        "kind": "ClusterIssuer",
        "name": "wildcard-letsencrypt-prod",
      },
      "secretName": "pennlabs-org-tls",
    },
  },
  Object {
    "apiVersion": "batch/v1beta1",
    "kind": "CronJob",
    "metadata": Object {
      "labels": Object {
        "name": "RELEASE_NAME-clear-expired-tokens",
        "release": "RELEASE_NAME",
      },
      "name": "RELEASE_NAME-clear-expired-tokens",
      "namespace": "default",
    },
    "spec": Object {
      "failedJobsHistoryLimit": 1,
      "jobTemplate": Object {
        "spec": Object {
          "template": Object {
            "spec": Object {
              "containers": Array [
                Object {
                  "command": Array [
                    "python3",
                    "manage.py",
                    "cleartokens",
                  ],
                  "env": Array [
                    Object {
                      "name": "GIT_SHA",
                      "value": "TAG_FROM_CI",
                    },
                  ],
                  "envFrom": Array [
                    Object {
                      "secretRef": Object {
                        "name": "platform",
                      },
                    },
                  ],
                  "image": "pennlabs/platform:TAG_FROM_CI",
                  "imagePullPolicy": "IfNotPresent",
                  "name": "worker",
                  "ports": Array [
                    Object {
                      "containerPort": 80,
                    },
                  ],
                },
              ],
              "restartPolicy": "Never",
            },
          },
        },
      },
      "schedule": "0 5 * * SUN",
      "successfulJobsHistoryLimit": 1,
    },
  },
]
`;
